src/main.rs

```diff
@@ -201,6 +232,112 @@ fn main() -> ! {
     }
 }
 
+fn check_and_process_control_can_frames(
+    modules: &mut ControlModules,
+    can_gateway: &mut CanGatewayModule,
+    debug_console: &mut DebugConsole,
+) -> Result<(), OxccError> {
+    // poll both control CAN FIFOs
+    for fifo in &[RxFifo::Fifo0, RxFifo::Fifo1] {
+        match can_gateway.control_can().receive(fifo) {
+            Ok(rx_frame) => {
+                modules.brake.process_rx_frame(&rx_frame, debug_console)?;
+                modules
+                    .throttle
+                    .process_rx_frame(&rx_frame, debug_console)?;
+                modules
+                    .steering
+                    .process_rx_frame(&rx_frame, debug_console)?;
+            }
+            Err(e) => {
+                // report all but BufferExhausted (no data)
+                if e != CanError::BufferExhausted {
+                    return Err(OxccError::from(e));
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn check_for_faults(
+    modules: &mut ControlModules,
+    can_gateway: &mut CanGatewayModule,
+    timer_ms: &MsTimer,
+    debug_console: &mut DebugConsole,
+) -> Result<(), OxccError> {
+    let maybe_fault = modules.brake.check_for_faults(timer_ms, debug_console)?;
+    if let Some(brake_fault) = maybe_fault {
+        can_gateway.publish_fault_report(brake_fault)?;
+    }
+
+    let maybe_fault = modules.throttle.check_for_faults(timer_ms, debug_console)?;
+    if let Some(throttle_fault) = maybe_fault {
+        can_gateway.publish_fault_report(throttle_fault)?;
+    }
+
+    let maybe_fault = modules.steering.check_for_faults(timer_ms, debug_console)?;
+    if let Some(steering_fault) = maybe_fault {
+        can_gateway.publish_fault_report(steering_fault)?;
+    }
+
+    Ok(())
+}
+
+fn publish_reports(
+    modules: &mut ControlModules,
+    can_gateway: &mut CanGatewayModule,
+) -> Result<(), OxccError> {
+    // attempt to publish them all, only report the first to fail
+    let mut result = Ok(());
+
+    if let Err(e) = can_gateway.publish_brake_report(modules.brake.supply_brake_report()) {
+        result = Err(OxccError::from(e));
+    }
+
+    if let Err(e) = can_gateway.publish_throttle_report(modules.throttle.supply_throttle_report()) {
+        result = Err(OxccError::from(e));
+    }
+
+    if let Err(e) = can_gateway.publish_steering_report(modules.steering.supply_steering_report()) {
+        result = Err(OxccError::from(e));
+    }
+
+    result
+}
+
+// TODO - this is just an example for now
+fn handle_error(
+    error: OxccError,
+    modules: &mut ControlModules,
+    can_gateway: &mut CanGatewayModule,
+    debug_console: &mut DebugConsole,
+    leds: &mut Leds,
+) {
+    let mut ignore_error: bool = false;
+
+    // it is typically to get timeout errors if the CAN bus is not active or
+    // there are no other nodes connected to it
+    if error == OxccError::CanBusTxTimeout {
```

Comment:
As this system becomes more complicated, it will likely become worthwhile to make a more action-local decision to ignore/not-propagate the precursor errors to this variant.  Limiting the spread of locally-remediable has the upside of reducing the temptation to do deeper global state inspection at error-handling time, which in turn can lead to subtle interconnection between seemingly unrelated aspects of state-tracking and error codes.

Another alternative to preventing error-handler tendril creep is to carry more decision-making data with the error objects themselves. e.g. I can imagine a future where certain CAN bus timeouts are completely ignorable, while others require some different action, and that differences in response can be captured or inferable from data in the error variant.

Reply:
Totally agree with you on that one. I've got that pushed down a little further in the stack now, with notes around it likely changing in the near future.

We definitely ought to do something in most cases.
One easy idea I had was to put the CAN bus in a passive mode, disable controls and ignore further enable commands until we can successfully transmit a frame or error bits are cleared/etc.
I'll run some other ideas by you on implementing parts of the error management once the main logic is in better shape.